### ISSUE TITLE
Consider using `final` on derived (component) classes

### ISSUE THREAD

**Issue by Kronos3** [nasa/fprime#3038]
| | |
|:---|:---|
|**_F´ Version_**| |
|**_Affected Component_**| |
---
## Feature Description

Add `final` to component implementations to allow compiler to remove one layer of indirection to port calls.

## Rationale

When calling a virtual function, the compiler will add an operation that checks the vtable for an override above it's own implementation. We can use `final` to "devirtualize" the CPP implementations at compile time.

For our use case `final` is perfect for our derived component classes. At the moment we are passing a `virtual` function pointer (input port call) to connect output -> input ports. This means that there are two levels of indirection in a port call:

1. Load the target input port call address (function pointer)
2. Check the vtable for an override (which does not exist)

Read more about it here https://llvm.org/devmtg/2016-11/Slides/Padlewski-DevirtualizationInLLVM.pdf

**Comment by LeStarch** [comment]
This makes sense to me!  Our components function as final classes anyway.

**Comment by LeStarch** [comment]
At one time, this would not have been desired behavior as there was a concept of using inheritance to make similar components that derived from a common interface.  However, we have converged on the concept of using FPP style interfaces as the preferred method resulting in distinct components that share the same shape.

All this is to say, from a development perspective and an F Prime/FPP assumption perspective, this is likely preferred regardless of the value performance improvement.  @timcanham @bocchino do you concur?

**Comment by celskeggs** [comment]
I'll also second this request, not just because of performance, but because it helps the C++ compiler detect missing port handlers sooner. If I compile just a single component, the build can succeed even if the component is missing a handler. (I won't hit an error until I construct an instance of the component in my topology.) If we include the final keyword by default, the missing handler will be caught very quickly when compiling the component.

**Comment by matt392code** [comment]
Proposed implementation approach for adding `final` to F Prime components:
1. Autocoder Modifications:
```cpp
// Current generated base class
class ${name}ComponentBase {
    virtual void ${port}_handler(NATIVE_INT_TYPE portNum, ${args}...);
};

// Modified to support final
class ${name}ComponentBase {
    // Mark virtual in base but allow final in implementation
    virtual void ${port}_handler(NATIVE_INT_TYPE portNum, ${args}...);
};

// Generated implementation class
class ${name} final : public ${name}ComponentBase {
    // Now marked final to enable devirtualization
    void ${port}_handler(NATIVE_INT_TYPE portNum, ${args}...) final override;
};
```
2. Build System Integration:
- Add CMake option to control finalization:
```cmake
option(FPRIME_ENABLE_FINAL "Enable final keyword on components" ON)
```
- Generate appropriate code based on option:
```cpp
#if FPRIME_ENABLE_FINAL
#define FPRIME_FINAL final
#else 
#define FPRIME_FINAL
#endif

class ${name} FPRIME_FINAL : public ${name}ComponentBase {
    // ...
};
```
3. Vtable Optimization Benefits:
- Single pointer dereference for port calls
- Better inlining opportunities
- Reduced code size since no vtable needed for final classes

4. Migration Strategy:
a. For new components:
- Automatically generate with final
- Full optimization benefits

b. For existing components:
- Optional CMake flag to enable gradually
- Allow manual annotation for specific components
- Maintain backwards compatibility

5. Implementation Constraints:
- Must maintain ABI compatibility for existing deployments
- Need to handle any existing derived components
- Consider impact on unit testing frameworks

**Comment by mprather** [comment]
I suggest dropping the build system integration for existing components. It's too much complexity for such a small detail that in many ways is super hard to actually realize (you're not going to flip that flag and suddenly get oodles of perf back). Stamp it out in the template and let devs figure out if they want/need to manually adjust existing components.

With `final override` addition to the class method, the general recommendation is to use only 1 word. In this case, simply `final` would provide the intent.  However, if final is specified at the class level, there is no need to be verbose at the method level b/c even if you specify `override` the method would be sealed. See https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c128-virtual-functions-should-specify-exactly-one-of-virtual-override-or-final

**Comment by matt392code** [comment]
Thanks for the feedback @mprather! I agree with simplifying the implementation by removing the build system integration. Here's the proposed change to the autocoder templates:

```cpp
// Generated implementation class
class ${name} final : public ${name}ComponentBase {
    void ${port}_handler(NATIVE_INT_TYPE portNum, ${args}...);
};
```

This would:
- Add `final` only at the class level for new components
- Skip the CMake configuration complexity
- Let developers manually modify existing components if desired
- Maintain clean, simple code generation

Does this align with what you had in mind? Once we agree on this approach, I can prepare the implementation.
```

**Comment by bocchino** [comment]
> All this is to say, from a development perspective and an F Prime/FPP assumption perspective, this is likely preferred regardless of the value performance improvement. @timcanham @bocchino do you concur?

Yes, I think this is a good idea. I think there are two steps here:
1. Add `final` to the component implementations in the F Prime core repo.
2. Add `final` to the component implementation skeletons generated by FPP.

**Comment by mprather** [comment]
@matt392code yes, your suggestion represents my suggestion. Rob has some additional feedback to help provide a more comprehensive update.

**Comment by matt392code** [comment]
Proposed implementation for adding the `final` keyword to F Prime components:
[fprime-final.txt](https://github.com/user-attachments/files/18342286/fprime-final.txt)
This implementation includes:

1. Base template modifications that keep the virtual methods in the base class
2. Implementation template changes to add the `final` keyword to component classes
3. Example of a generated component showing how it would look in practice
4. FPP template updates to support the `final` keyword generation

Key points about this implementation:

1. The base class maintains virtual methods to define the component interface
2. The implementation class is marked `final` to prevent inheritance
3. Port handlers use `override` but don't need `final` since the class is final
4. Templates are updated to automatically generate the `final` keyword for new components
